src/Microsoft.AspNet.Mvc/MvcServices.cs

```diff
@@ -81,10 +82,16 @@ public static IEnumerable<IServiceDescriptor> GetDefaultServices(IConfiguration
             yield return describe.Transient<IFilterProvider, DefaultFilterProvider>();
 
             yield return describe.Transient<FormatFilter, FormatFilter>();
+
             // Dataflow - ModelBinding, Validation and Formatting
-            // The DataAnnotationsModelMetadataProvider does significant caching of reflection/attributes
-            // and thus needs to be singleton. 
-            yield return describe.Singleton<IModelMetadataProvider, DataAnnotationsModelMetadataProvider>();
+            //
+            // The DefaultModelMetadataProvider does significant caching and should be a singleton.
+            yield return describe.Singleton<ModelBinding.IModelMetadataProvider, DefaultModelMetadataProvider>();
+            yield return describe.Transient<ModelBinding.Metadata.ICompositeMetadataDetailsProvider>(services =>
+            {
+                var options = services.GetRequiredService<IOptions<MvcOptions>>().Options;
```

Comment:
interesting what was the downside of doing this in the constructor ( which has been the pattern we've followed so far ).
Exposing all options instead of exactly what is needed?